spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandPruning.scala

```diff
@@ -0,0 +1,150 @@
+import org.apache.spark.sql.catalyst.trees.TreePattern.SORT
+import org.apache.spark.sql.connector.read.SupportsRuntimeFiltering
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+/**
+ * A rule that adds a runtime filter for row-level commands.
+ *
+ * Note that only group-based rewrite plans (i.e. ReplaceData) are taken into account.
+ * Row-based rewrite plans are subject to usual runtime filtering.
+ */
+case class RowLevelCommandPruning(spark: SparkSession) extends Rule[LogicalPlan] with PredicateHelper {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    // apply special dynamic filtering only for plans that don't support deltas
+    case RewrittenRowLevelCommand(
+        command: RowLevelCommand,
+        DataSourceV2ScanRelation(_, scan: SupportsRuntimeFiltering, _),
+        rewritePlan: ReplaceData) if conf.dynamicPartitionPruningEnabled =>
+
+      // use reference equality to find exactly the required scan relations
+      val newRewritePlan = rewritePlan transformUp {
+        case r: DataSourceV2ScanRelation if r.scan eq scan =>
+          val pruningKeys = ExtendedV2ExpressionUtils.resolveRefs[Attribute](scan.filterAttributes, r)
+          val dynamicPruningCond = buildDynamicPruningCondition(r, command, pruningKeys)
+          val filter = Filter(dynamicPruningCond, r)
+          // always optimize dynamic filtering subqueries for row-level commands as it is important
+          // to rewrite introduced predicates as joins because Spark recently stopped optimizing
+          // dynamic subqueries to facilitate broadcast reuse
+          optimizeSubquery(filter)
+      }
+      command.withNewRewritePlan(newRewritePlan)
+  }
+
+  private def buildDynamicPruningCondition(
+      relation: DataSourceV2ScanRelation,
+      command: RowLevelCommand,
+      pruningKeys: Seq[Attribute]): Expression = {
+
+    // construct a filtering plan with the original scan relation
+    val cond = command.condition.getOrElse(Literal.TrueLiteral)
+    val matchingRowsPlan = command match {
+      case m: MergeIntoTable =>
+        Join(relation, m.sourceTable, LeftSemi, Some(cond), JoinHint.NONE)
+
+      case u: UpdateTable =>
+        // UPDATEs with subqueries may be rewritten using a UNION with two identical scan relations
+        // each scan relation will get its own dynamic filter that will be shared during execution
+        // the analyzer will assign different expr IDs for each scan relation output attributes
+        // that's why the condition may refer to invalid attr expr IDs and must be transformed
+        val attrMap = AttributeMap(u.table.output.zip(relation.output))
+        val transformedCond = cond transform {
+          case attr: AttributeReference if attrMap.contains(attr) => attrMap(attr)
+        }
+        Filter(transformedCond, relation)
+
+      case _ =>
+        Filter(cond, relation)
```

Comment:
It makes sense to me to use the original delete condition here for the dynamic pruning.

I would prefer to make the command matching more explicit though, rather than doing this for anything that doesn't match `MergeIntoTable` or `UpdateTable`.

Reply:
Good call.

Reply:
Fixed.